versioned_hdf5/slicetools.pyx

```diff
@@ -234,6 +238,125 @@ cpdef build_data_dict(dcpl, raw_data_name: str):
+        # FIXME Error getting virtual sources on empty virtual dataset
+        # https://github.com/h5py/h5py/issues/1660
+        if dcpl.get_layout() != h5d.VIRTUAL or 0 in shape:
+            return slab_indices, slab_offsets
+
+        dcpl_id: hid_t = dcpl.id
+        virtual_count: size_t = dcpl.get_virtual_count()
+
+        with nogil:
+            # Iterate over all chunks except full ones
+            for chunk_i in range(virtual_count):
+                # Calculate offset of the chunk, IN BYTES, relative to the start of the
+                # flattened buffer of slab_indices and slab_offsets
+                vspace_id = H5Pget_virtual_vspace(dcpl_id, chunk_i)
+                if vspace_id == H5I_INVALID_HID:
+                    raise HDF5Error()
+                try:
+                    vspace_type = H5Sget_select_type(vspace_id)
+                    if vspace_type != H5S_sel_type.H5S_SEL_HYPERSLABS:
+                        raise ValueError(f"Unexpected {vspace_type=}")
+
+                    # Get chunk index in the virtual dataset
+                    if H5Sget_regular_hyperslab(
+                        vspace_id,
+                        h5_start,
+                        scratch,  # IGNORE - stride (always 1)
+                        scratch,  # IGNORE - count
+                        scratch,  # IGNORE - block (typically, but not always, 1)
+                    ) < 0:
+                        raise HDF5Error()
+                finally:
+                    if H5Sclose(vspace_id) < 0:
+                        raise HDF5Error()
+
+                indices_offset: ssize_t = 0
+                for j in range(ndim):
+                    chunk_idx = h5_start[j] // chunk_size_arr[j]
+                    if h5_start[j] % chunk_size_arr[j] > 0:
+                        raise ValueError("Misaligned chunk")
+                    if chunk_idx >= nchunks_arr[j]:
+                        raise ValueError("chunk_idx too high")
+                    indices_offset += chunk_idx * indices_strides[j]
+
+                # Calculate offset of the chunk, IN POINTS,
+                # relative to the start of raw_data along axis 0
+                srcspace_id = H5Pget_virtual_srcspace(dcpl_id, chunk_i)
+                if srcspace_id == H5I_INVALID_HID:
+                    raise HDF5Error()
+                try:
+                    srcspace_type = H5Sget_select_type(srcspace_id)
+                    if srcspace_type != H5S_sel_type.H5S_SEL_HYPERSLABS:
+                        raise ValueError(f"Unexpected {srcspace_type=}")
+
+                    # Get coordinates in raw_data
+                    if H5Sget_regular_hyperslab(
+                        srcspace_id,
+                        h5_start,  # Axis 0 = slab_offset; other axes = always 0
+                        scratch,   # IGNORE - stride (always 1)
+                        scratch,   # IGNORE - count
+                        scratch,   # IGNORE - block (typically, but not always, 1)
```

Comment:
Inconsequential as long as raw_data is contiguous, which is always the case. Chunk size can be trivially calculated from the global chunk size plus a cut to shape to be applied to the edge chunks